File: packages/ts/core/src/models/request.ts
import { z } from "zod";

import { MessageSchema } from "./messages.js";

/** Reasoning effort hint for models that support it. */
export const ReasoningEffortSchema = z.enum(["low", "medium", "high"]);
export type ReasoningEffort = z.infer<typeof ReasoningEffortSchema>;

/** A tool the model may call. `parameters` is an arbitrary JSON Schema object. */
export const ToolDefinitionSchema = z.object({
  name: z.string(),
  description: z.string(),
  parameters: z.record(z.string(), z.unknown()),
});
export type ToolDefinition = z.infer<typeof ToolDefinitionSchema>;

/** How the model should select tools. */
export const ToolChoiceSchema = z.union([
  z.enum(["auto", "none", "required"]),
  z.object({ type: z.literal("tool"), name: z.string() }),
]);
export type ToolChoice = z.infer<typeof ToolChoiceSchema>;

/** The provider-agnostic inbound request every adapter receives. */
export const ChatRequestSchema = z.object({
  model: z.string(),
  messages: z.array(MessageSchema),
  temperature: z.number().min(0).max(2).optional(),
  maxTokens: z.number().int().positive().optional(),
  reasoningEffort: ReasoningEffortSchema.optional(),
  tools: z.array(ToolDefinitionSchema).optional(),
  toolChoice: ToolChoiceSchema.optional(),
  responseSchema: z.record(z.string(), z.unknown()).optional(),
});
export type ChatRequest = z.infer<typeof ChatRequestSchema>;

File: packages/py/llmorch-core/src/llmorch_core/models/request.py
"""Inbound request models — twin of packages/ts/core/src/models/request.ts."""

from typing import Any, Literal

from pydantic import Field

from .base import CamelModel
from .messages import Message

ReasoningEffort = Literal["low", "medium", "high"]


class ToolDefinition(CamelModel):
    name: str
    description: str
    parameters: dict[str, Any]


class NamedToolChoice(CamelModel):
    type: Literal["tool"]
    name: str


ToolChoice = Literal["auto", "none", "required"] | NamedToolChoice


class ChatRequest(CamelModel):
    model: str
    messages: list[Message]
    temperature: float | None = Field(default=None, ge=0, le=2)
    max_tokens: int | None = Field(default=None, alias="maxTokens", gt=0)
    reasoning_effort: ReasoningEffort | None = Field(default=None, alias="reasoningEffort")
    tools: list[ToolDefinition] | None = None
    tool_choice: ToolChoice | None = Field(default=None, alias="toolChoice")
    response_schema: dict[str, Any] | None = Field(default=None, alias="responseSchema")
